component/async/component_test.go

```diff
@@ -206,6 +209,41 @@ func TestRun_ProcessError_WithNackError(t *testing.T) {
 	assert.Equal(t, 1, builder.proc.execs)
 }
 
+// TestRun_ParallelProcessError_WithNackError expects a PROC ERROR
+// from an error producing processor
+// but due to the concurrency setting, it will continue processing other messages
+func TestRun_ParallelProcessError_WithNackError(t *testing.T) {
+	builder := proxyBuilder{
+		proc: mockProcessor{errReturn: true},
+		cnr: mockConsumer{
+			chMsg: make(chan Message, 10),
+			chErr: make(chan error, 10),
+		},
+		fs:          NackStrategy,
+		concurrency: 10,
+	}
+
+	ctx, cnl := context.WithCancel(context.Background())
+	builder.cnr.chMsg <- &mockMessage{ctx: ctx, nackError: true}
+	ch := make(chan bool)
+	go func() {
+		assert.NoError(t, run(ctx, t, &builder))
```

Comment:
```go
func TestGoroutine1(t *testing.T) {
	go func() {
		time.Sleep(time.Second)
		t.FailNow()
	}()
}
```

```go
func TestGoroutine2(t *testing.T) {
	go func() {
		time.Sleep(time.Second)
		t.FailNow()
	}()
	time.Sleep(2 * time.Second)
}
```

Output:
```
--- PASS: TestGoroutine1 (0.00s)
--- FAIL: TestGoroutine2 (2.00s)
```


If this goroutine exits after the test, the test will pass no matter what.
Is the `time.Sleep` after that here to bypass that? We have no guarantee that the test will always pass - especially if running other tests in parallel, or using a slowing CPU.

We could maybe use the `ch` here, to make sure the test finished: 

```go
ch := make(chan error)

go func() {
 	ch <- run(ctx, t, &builder)
}()

select {
case _, ok := <-builder.cnr.chErr:
 	if ok {
 		assert.Fail(t, "we don't expect an error , given our ack failure strategy setup")
	}
case err := <-ch:
       assert.NoError(t, err)
}
```

In other words, moving `<-ch` out of the `default` case, and as a clause of the select.

Reply:
Note that `case _, ok := <-builder.cnr.chErr:` will also be reached when we `close(builder.cnr.chErr)`, with `ok == false`.

I'm assuming this is intentional, since there's an `ok` check afterwards.

Reply:
As mentioned early, it's just a copy paste from a previous test, which I'm assuming to be working fine. 10ms should be enough, unless of course the machine running the tests has a cluttered CPU. Let's see if other people think we should make it more resilient, I'm happy to do it.